core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -0,0 +1,496 @@
+              "[%s] Unable to parse the server CA certificate for the Cloud SQL instance.",
+              connectionName),
+          ex);
+    }
+
+    return new Metadata(ipAddrs, instanceCaCertificate);
+  }
+
+  /**
+   * Uses the Cloud SQL Admin API to create an ephemeral SSL certificate that is authenticated to
+   * connect the Cloud SQL instance for 60 minutes.
+   */
+  private Certificate fetchEphemeralCertificate() {
+    // Format the public key into a PEM encoded Certificate.
+    StringBuilder publicKeyPemBuilder = new StringBuilder();
+    publicKeyPemBuilder.append("-----BEGIN RSA PUBLIC KEY-----\n");
+    publicKeyPemBuilder.append(
+        Base64.getEncoder()
+            .encodeToString(keyPair.getPublic().getEncoded())
+            .replaceAll("(.{64})", "$1\n"));
+    publicKeyPemBuilder.append("\n");
+    publicKeyPemBuilder.append("-----END RSA PUBLIC KEY-----\n");
+
+    // Use the SQL Admin API to create a new ephemeral certificate.
+    SslCertsCreateEphemeralRequest request =
+        new SslCertsCreateEphemeralRequest().setPublicKey(publicKeyPemBuilder.toString());
+    SslCert response;
+    try {
+      response = apiClient.sslCerts().createEphemeral(projectId, instanceId, request).execute();
+    } catch (IOException ex) {
+      throw addExceptionContext(
+          ex,
+          String.format(
+              "[%s] Failed to create ephemeral certificate for the Cloud SQL instance.",
+              connectionName));
+    }
+
+    // Parse the certificate from the response.
+    Certificate ephemeralCertificate;
+    try {
+      byte[] certBytes = response.getCert().getBytes(StandardCharsets.UTF_8);
+      ephemeralCertificate =
+          CertificateFactory.getInstance("X.509")
+              .generateCertificate(new ByteArrayInputStream(certBytes));
+    } catch (CertificateException ex) {
+      throw new RuntimeException(
+          String.format(
+              "[%s] Unable to parse the ephemeral certificate for the Cloud SQL instance.",
+              connectionName),
+          ex);
+    }
+    return ephemeralCertificate;
+  }
+
+  // Schedules task to be executed once the provided futures are complete.
+  private <T> ListenableFuture<T> whenComplete(Callable<T> task, ListenableFuture<?>... futures) {
+    SettableFuture<T> taskFuture = SettableFuture.create();
+
+    // Create a countDown for all Futures to complete.
+    AtomicInteger countDown = new AtomicInteger(futures.length);
+
+    // Trigger the task when all futures are complete.
+    Runnable runOutput =
+        () -> {
+          if (countDown.decrementAndGet() == 0) {
+            taskFuture.setFuture(executor.submit(task));
+          }
+        };
+    for (ListenableFuture<?> future : futures) {
+      future.addListener(runOutput, executor);
+    }
+
+    return taskFuture;
+  }
+
+  // Returns the inner future from a nested future, or else the exception it encountered.
+  private <T> ListenableFuture<T> extractNestedFuture(
```

Comment:
This can be static to indicate that it's a utility that isn't tied to this instance.